---
output:
  md_document:
    variant: markdown_github
bibliography: library.bib
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

# SCORPIUS

[![Build Status](https://travis-ci.org/rcannood/SCORPIUS.png?branch=master)](https://travis-ci.org/rcannood/SCORPIUS)

**SCORPIUS an unsupervised approach for inferring developmental chronologies from single-cell
RNA sequencing data.** In comparison to similar approaches, it has three main advantages:

* **It accurately reconstructs trajectories for a wide variety of dynamic 
  cellular processes.** The performance was evaluated using a new, quantitative 
  evaluation pipeline, comparing the performance of current state-of-the-art techniques
  on 10 publicly available single-cell RNA sequencing datasets. 
  
* **It automatically identifies marker genes, speeding up knowledge discovery.**
  
* **It is fully unsupervised.** Prior knowledge of the relevant marker genes or 
  cellular states of individual cells is not required.
  
## Installing SCORPIUS

You can install the latest version from github with

```R
devtools::install_github("rcannood/SCORPIUS")
```

<!--
You can install:

* the latest released version from CRAN with

    ```R
    install.packages("SCORPIUS")
    ```

* the latest development version from github with

    ```R
    devtools::install_github("rcannood/SCORPIUS")
    ```
-->

If you encounter a clear bug, please file a minimal reproducible example on [github](https://github.com/rcannood/SCORPIUS/issues). 


## Learning SCORPIUS

To get started, read the notes below or read the ~~[intro vignette](): `vignette("introduction", package="SCORPIUS")`~~ (Under construction).

More vignettes with more elaborate examples are also available: 

* ~~[Investigating differentiating dendritic cell progenitors](): `vignette("ginhoux", package="SCORPIUS")`~~ (Under construction)
* [Inferring trajectories from simulated data](vignettes/simulated-data.md): `vignette("simulated-data", package="SCORPIUS")`

## Minimal example
This section describes the main workflow of SCORPIUS without going in depth in the R code. For a more detailed explanation, see the `ginhoux` vignette.

To start using SCORPIUS, simply write:
```{r libraries, message=FALSE}
library(SCORPIUS)
```


The `ginhoux` dataset [See @Schlitzer2015] contains 248 cell progenitors in one of three cellular cellular states: MDP, CDP or PreDC.
```{r load ginhoux}
data(ginhoux)
expression <- ginhoux$expression
group.name <- ginhoux$sample.info$group.name
```


With the following code, SCORPIUS reduces the dimensionality of the dataset and provides a visual overview of the dataset. 
In this plot, cells that are similar in terms of expression values will be placed closer together than cells with dissimilar expression values.

```{r reduce dimensionality}
dist <- correlation.distance(expression)

filt <- outlier.filter(dist)
expression <- expression[filt, ]
group.name <- group.name[filt]
dist <- dist[filt, filt]

space <- reduce.dimensionality(dist, ndim=2)
draw.trajectory.plot(space, group.name)
```


To infer and visualise a trajectory through these cells, run:
```{r infer trajectory}
traj <- infer.trajectory(space)
draw.trajectory.plot(space, group.name, traj$final.path)
```


Finally, to identify and visualise candidate marker genes, execute the following code:
```{r find tafs sequentially, eval=F}
tafs <- find.trajectory.aligned.features(expression, traj$time)
expr.tafs <- tafs$smooth.x[,tafs$tafs]
modules <- extract.modules(expr.tafs)
draw.trajectory.heatmap(expr.tafs, traj$time, group.name, modules)
```
```{r find tafs, echo=F, message=F}
tafs <- find.trajectory.aligned.features(expression, traj$time, parallel=8)
expr.tafs <- tafs$smooth.x[,tafs$tafs]
modules <- extract.modules(expr.tafs)
draw.trajectory.heatmap(expr.tafs, traj$time, group.name, modules)
```


## Related approaches

* [Monocle](https://bioconductor.org/packages/release/bioc/html/monocle.html)
* [Waterfall](http://dx.doi.org/10.1016/j.stem.2015.07.013)
* [Embeddr](https://github.com/kieranrcampbell/embeddr)


## References
