---
title: "Trajectory inference from simulated data"
author: "Robrecht Cannoodt"
date: "2016-01-22"
output: 
  rmarkdown::html_vignette: 
    fig_height: 6
    fig_width: 6
  md_document:
    variant: markdown_github
vignette: >
  %\VignetteIndexEntry{Trajectory inference from simulated data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<!-- github markdown built using 
render("vignettes/simulated-data.Rmd", output_format = "md_document")
-->

```{r, echo=F}
set.seed(4)
```

In this vignette, SCORPIUS is used to infer a trajectory through cells in artificial single-cell RNA-seq data.
Note that the dataset is generated in a very naive manner and is only meant to be used for demonstration purposes, not for evaluating trajectory inference methods.


## Simulate expression data
Expression values for 384 cells and 500 genes is generated as follows.
```{r message=FALSE}
library(SCORPIUS)
dataset <- generate.dataset(type="poly", num.genes=500, num.samples=384, num.groups=4)
```

The resulting dataset is a list containing a matrix named `expression` and a data frame named `sample.info`.

`expression` is a 384-by-500 matrix containing the expression values of all the cells and all the genes.
```{r}
dataset$expression[1:6, 1:6]
```

`sample.info` is a data frame with the metadata of the cells, containing only the group each cell belongs to.
```{r}
head(dataset$sample.info)
```

In order to infer a trajectory through this data, SCORPIUS first reduces the dimensionality of the dataset.


## Reduce dimensionality of the dataset
SCORPIUS uses classical Torgerson multi-dimensional scaling to reduce the dataset to three dimensions.
In short, this technique attempts to place the cells in a space such that the distance between any
two points in that space approximates the original distance between the two cells as well as possible.

The distance between any two samples is defined as their correlation distance, namely `1 - (cor(x, y)+1)/2`. 
The distance matrix is calculated as follows:
```{r}
expression <- dataset$expression
group.name <- dataset$sample.info$group.name
dist <- correlation.distance(expression)
```

`dist` is a 384-by-384 matrix, with values ranging from 0 to 1.
```{r}
dim(dist)
plot(density(dist))
```


The reduced space is constructed as follows:
```{r}
space <- reduce.dimensionality(dist)
```


The new space is a 384-by-3 matrix, and can be visualised as follows:
```{r}
draw.trajectory.plot(space)
```


Looking at this plot, it seems that the cells in this dataset are involved in a dynamic process. 

In addition, if a property of the cells (e.g. cell type) is known, it can be used to colour the plot using the `progression.group` parameter.

In this case, the underlying groups of each cell were also given:
```{r}
draw.trajectory.plot(space, progression.group = group.name)
```


## Inferring a trajectory through the cells
The main goal of SCORPIUS is to infer a trajectory through the cells, and orden the cells according to the inferred timeline.

SCORPIUS infers a trajectory through several intermediate steps, which are all executed as follows:
```{r}
traj <- infer.trajectory(space)
```


The result is a list containing the final trajectory `path` and the inferred timeline for each sample `time`.

The trajectory can be visualised with respect to the samples by passing it to `draw.trajectory.plot`:
```{r}
draw.trajectory.plot(space, progression.group = group.name, path = traj$path)
```


## Finding candidate marker genes
We search for genes whose expression is seems to be a function of the trajectory timeline that was inferred, 
as such genes might be good candidate marker genes for the dynamic process that is being investigated.

```{r find tafs}
tafs <- find.trajectory.aligned.features(expression, traj$time, verbose=F)
```

Again, the output is a list containing several values:

* `tafs`: The names of genes which are marked as trajectory aligned.

* `p.values`: A data frame containing all the genes and corresponding q-values.

* `smooth.x`: The smoothed expression data used in findingn the TAFs.

To visualise the expression of the selected TAFs, use the `draw.trajectory.heatmap` function.
```{r visualise tafs}
expr.tafs <- tafs$smooth.x[,tafs$tafs]
draw.trajectory.heatmap(expr.tafs, traj$time, group.name)
```

Finally, the TAFs can also be grouped into modules as follows: 
```{r moduled tafs}
modules <- extract.modules(expr.tafs)
draw.trajectory.heatmap(expr.tafs, traj$time, group.name, modules)
```
